src/main/host/thread_ptrace.c

```diff
@@ -386,42 +479,165 @@ static pid_t _waitpid_spin(pid_t pid, int *wstatus, int options) {
+        // if ptrace events are a special case.
+        int wstatus;
+        pid_t pid = waitpid(thread->base.nativeTid, &wstatus, WNOHANG);
+        if (pid < 0) {
+            error("waitpid: %s", strerror(pid));
+            abort();
+        }
+        if (pid != 0) {
+            debug("Got ptrace stop");
+            StopReason ptraceStopReason = _getStopReason(wstatus);
+
+            // Pre-emptively save registers here.  TODO: do this lazily, since
+            // we won't always need them. Doing it here makes it easier to
+            // ensure we get the right value for rax though, since we know what
+            // kind of ptrace stop just happened.
+            if (ptrace(PTRACE_GETREGS, thread->base.nativeTid, 0, &thread->regs.value) < 0) {
+                error("ptrace: %s", g_strerror(errno));
+                abort();
+            }
+            if (ptraceStopReason.type == STOPREASON_SYSCALL) {
+                debug("Saving syscall rip");
+                thread->regs.value.rax = thread->regs.value.orig_rax;
+                thread->syscall_rip = thread->regs.value.rip - sizeof(SYSCALL_INSTRUCTION);
+            }
+            thread->regs.valid = true;
+            thread->regs.dirty = false;
+
+            if (thread->enableIpc &&
+                shimevent_tryRecvEventFromPlugin(
+                    _threadptrace_ipcData(thread), &event_stop.shim_event) == 0) {
+                // The plugin finished sending an event after our previous
+                // attempt to receive it, and then hit a ptrace-stop.  We need
+                // to handle the sent-event first, and buffer the ptrace-stop
+                // to be handled later.  In particular, the ptrace-stop could
+                // be a blocking futex syscall on the shim IPC control
+                // structures; if we try to execute it before responding to the
+                // shim event, we could deadlock.
+                debug("Buffering ptrace-stop while handling shim event");
+                thread->ipc_syscall.stopped = true;
+                thread->ipc_syscall.pendingStop = ptraceStopReason;
+                thread->ipc_syscall.havePendingStop = true;
+                return event_stop;
+            }
+            return ptraceStopReason;
+        }
+    };
+}
 
-    StopReason reason = _getStopReason(wstatus);
+static void _threadptrace_nextChildState(ThreadPtrace* thread) {
+    StopReason reason = _threadptrace_hybridSpin(thread);
     _threadptrace_updateChildState(thread, reason);
 }
 
 pid_t threadptrace_run(Thread* base, gchar** argv, gchar** envv) {
     ThreadPtrace* thread = _threadToThreadPtrace(base);
 
-    thread->base.nativeTid = _threadptrace_fork_exec(argv[0], argv, envv);
+    /* set the env for the child */
+    gchar** myenvv = g_strdupv(envv);
+
+    if (thread->enableIpc) {
+        ShMemBlockSerialized ipcBlkSerial = shmemallocator_globalBlockSerialize(&thread->ipcBlk);
```

Comment:
Does it make any sense to (optionally) set up this block in the parent class? Both subclasses currently implement this behavior.